### PR TITLE
added winston logger, added request info and error info loggers

### PR DESCRIPTION
**Pull request by alexandrali**

### CODE REVIEW

File: src/index.ts [hunk 1]
```diff
@@ -2,6 +2,14 @@ import express from 'express';
 import config from './config';
 import loaders from './loaders';
 
+process.on('uncaughtException', err => {
+  throw new Error(err.message);
```
Comment (titivuk) [inline src/index.ts]: Add logging. What's a point of catching it and throwing. The same behaviour if you delete it?

File: src/index.ts [hunk 2]
```diff
@@ -2,6 +2,14 @@ import express from 'express';
...
+});
+
+process.on('unhandledRejection', (reason, promise) => {
+  throw new Error(`Unhandled Rejection at: ${promise}. Reason: ${reason}`);
```
Comment (titivuk) [inline src/index.ts]: add logging. What's a point of catching it and throwing. The same behaviour if you delete it?

File: src/loaders/express.ts [hunk 3]
```diff
@@ -2,14 +2,52 @@ import express, {Response, Request, NextFunction} from 'express';
 import {RESPONSE_MESSAGES} from '../config/messages';
 import usersRouter from '../routers/users-routes';
 import groupsRouter from '../routers/groups-routes';
+import logger from '../config/logger';
+
+function getRequestData(req: Request) {
+  const {method, url} = req;
```
Comment (titivuk) [inline src/loaders/express.ts]: you can also do params, queryParams, headers

File: src/config/logger.ts [hunk 4]
```diff
@@ -0,0 +1,21 @@
+import winston from 'winston';
+
+const logger = winston.createLogger({
+  level: 'info',
+  format: winston.format.combine(
```
Comment (titivuk) [inline src/config/logger.ts:5]: I would use structured logging format. It's basically log everything as JSON object. Structured logs can be parsed by 3rd party logs system: AWS CloudWatch, for example. It allows you to filter logs by some nested fields
```suggestion
  format: winston.format.combine(
    winston.format.timestamp(),
    winston.format.json()
  ),
```

Comment (titivuk) [reply src/config/logger.ts:5]: In this case, I'm not sure that you need abstractions like requestSent and requestError. You can use logger directly

Comment (titivuk) [reply src/config/logger.ts:5]: example, logging all incoming requests

`logger.info(`Invoking request ${method} ${url}`, {method, url, body});`

output

```json
{
  "body": { "age": 30, "login": "llllll", "password": "********" },
  "level": "info",
  "message": "Invoking request POST /users",
  "method": "POST",
  "timestamp": "2023-03-06T09:43:14.778Z",
  "url": "/users"
}
```

Comment (titivuk) [reply src/config/logger.ts:5]: as I said, 3rd party logging systems understand json, and, for exmaple, you can apply filtering on json fields (find all logs where body.age = 30)

File: src/index.ts (hunk 1 above)
Comment (alexandrali) [reply src/index.ts]: added log instear of error throw

File: src/index.ts (hunk 2 above)
Comment (alexandrali) [reply src/index.ts]: added log instear of error throw

File: src/config/logger.ts (hunk 4 above)
Comment (alexandrali) [reply src/config/logger.ts:5]: used json format instead of string

**Review by titivuk — APPROVED** [review]

**Pull request merged**